### ISSUE TITLE
provide dependency injection for schema serialization

### ISSUE THREAD

**Issue by jone** [plone/plone.restapi#38]
The schema serialization is currently quite hard-coded for the standard fields.
When having custom fields, it is only possible to provide serialization by copying the serialization function and adapting it.

I'd like to change the serialization mechanism so that I'm able to change how a certain field value is represented by registering an adapter and easily provide custom serializer for my custom fields.

Maybe something like that:
```python
class IFieldSerializer(Interface):
    """The field serializer multi adapter serializes the field value into
    a JSON compatible python data. """

    def __init__(context, request, field):
        """Adapts context, request and the field.
        """

    def serialize():
        """Returns JSON compatible python data.
        """
```

Although the request is not really relevant here, I usually like multi adapters to also adapt the request so that I'm able to easily register a more specific adapter by using a (project-) request-layer as discriminator. Without adapting the request I'm more often forced to customize with an `overrides.zcml`, which I do not like. This is my personal taste and experience though :wink: 

Some questions:

1. Is there already some serialization mechanism which produces JSON-compatible output (!) for dexterity?
2. Where should this go? Into `plone.restapi` or into another package?
3. Any comments about my approach?

**Comment by jone** [comment]
Should this maybe be a two-way solution so that we can prepare data for updating / creating?

**Comment by tisto** [comment]
Please note that the schema serialization was just a proof-of-concept. That was never meant to be the final code.

Your questions:

1) Not that I would know.
2) Could go into plone.restapi first, could be refactored into a separate package as well.
3) Looks ok to me. Adapters always work. ;)

**Comment by jone** [comment]
Ok, thanks. I'll make a PR :wink:

**Comment by lukasgraf** [comment]
Looks good to me :)

> Is there already some serialization mechanism which produces JSON-compatible output (!) for dexterity?

I wouldn't know of one that does exactly this, but:
- [`plone.supermodel`](https://github.com/plone/plone.supermodel) provides XML import and export for **schema interfaces** based on `zope.schema` fields
- [`plone.rfc822`](https://github.com/plone/plone.rfc822) does marshalling of content objects described by `zope.schema` fields from and to RFC (2)822 messages (E-Mail / HTTP-Messages).

So even though they do something different, it still could be worth a look for some tricky cases, maybe we can pick up some odd bits and pieces from these packages.

**Comment by lukasgraf** [comment]
I think the `IFieldSerializer` adapter and your choice of discriminators is the way to go, but at some later stage we might need another layer of abstraction:

We'll have use cases where we'll want to serialize objects that aren't Plone content types (think SQL backed content for example). When using SQLAlchemy the `Column` objects could play a similar role to fields, in that the describe the data type, but I could imagine cases where you don't even have a field to adapt - you just know you're dealing with a datetime, and you want to be able to re-use the conversion to a JSON-compatible ISO format instead of writing it yourself.

So maybe at some later stage, we'd need some concept like plain `Converter`s, where a `DateTimeFieldSerializer` would delegate to a `DateTimeConverter` instead of doing the conversion itself.

**Comment by buchi** [comment]
I've got two suggestions of cosmetic nature:
* `IFieldSerializer` should adapt the field first as it is the main thing that is adapted:
  ```def __init__(self, field, context, request)```
* Do serialization in the `__call__()` method instead of a `serialize()`method. This reduces verbosity in code `serializer.serialize()` vs. `serializer()`

I've started implementing deserialization in #41 following this approach.
